### ISSUE TITLE
Fatal error: Declaration of Codeception\Test\Test::run(?PHPUnit\Framework\TestResult $result = NULL) must be compatible with PHPUnit\Framework\Test::run(?PHPUnit\Framework\TestResult $result = NULL): PHPUnit\Framework\TestResult in /var/www/html/vendor/codeception/codeception/src/Codeception/Test/Test.php on line 18

### ISSUE THREAD

**Issue by natenolting** [Codeception/Codeception#5523]
#### What are you trying to achieve?

Split tests with `codeception/robo-paracept` or run failed tests `codecept run -g failed` but `codecept run unit` will run without this error.

#### What do you get instead?

> Provide console output if related. Use `-vvv` mode for more details.

```bash
Fatal error: Declaration of Codeception\Test\Test::run(?PHPUnit\Framework\TestResult $result = NULL) must be compatible with PHPUnit\Framework\Test::run(?PHPUnit\Framework\TestResult $result = NULL): PHPUnit\Framework\TestResult in /var/www/html/vendor/codeception/codeception/src/Codeception/Test/Test.php on line 18

A PHP Error was encountered

Severity: Compile ErrorMessage:  Declaration of Codeception\Test\Test::run(?PHPUnit\Framework\TestResult $result = NULL) must be compatible with PHPUnit\Framework\Test::run(?PHPUnit\Framework\TestResult $result = NULL): PHPUnit\Framework\TestResultFilename: /var/www/html/vendor/codeception/codeception/src/Codeception/Test/Test.phpLine Number: 18

Backtrace:




```
Note: No backtrace is given.

> Provide test source code if related

### Details

* Codeception version: 3.0.1
* PHP Version:
```
PHP 7.2.17 (cli) (built: Apr  6 2019 03:27:44) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.17, Copyright (c) 1999-2018, by Zend Technologies
    with blackfire v1.25.0~linux-x64-non_zts72, https://blackfire.io, by Blackfire
```
* Operating System: Debian GNU/Linux 9 (stretch)
* Installation type: Composer
* List of installed packages (`composer show`)
```
anahkiasen/underscore-php              2.0.0              A redacted port of Underscore.js for PHP
behat/gherkin                          v4.6.0             Gherkin DSL parser for PHP 5.3
brainsware/php-markdown-extra-extended 0.1.0              PHP Markdown & Extra
brandonwamboldt/utilphp                1.0.5              util.php is a collection of useful functions and snip...
codeception/c3                         2.4.1              CodeCoverage collector for Codeception
codeception/codeception                3.0.1              BDD-style testing framework
codeception/phpunit-wrapper            8.0.4              PHPUnit classes used by Codeception
codeception/robo-paracept              0.4.2              Codeception Parallel Execution Tasks via Robo Task Ru...
codeception/stub                       2.1.0              Flexible Stub wrapper for PHPUnit's Mock Builder
codeigniter/framework                  3.1.10             The CodeIgniter framework
consolidation/annotated-command        2.12.0             Initialize Symfony Console commands from annotated co...
consolidation/config                   1.2.1              Provide configuration services for a commandline tool.
consolidation/log                      1.1.1              Improved Psr-3 / Psr\Log logger based on Symfony Cons...
consolidation/output-formatters        3.4.1              Format text by applying transformations provided by p...
consolidation/robo                     1.4.9              Modern task runner
consolidation/self-update              1.1.5              Provides a self:update command for Symfony Console ap...
container-interop/container-interop    1.2.0              Promoting the interoperability of container objects (...
crazyinventor/mailtrapper              master master     
dflydev/dot-access-data                v1.1.0             Given a deep data structure, access data by dot notat...
doctrine/inflector                     v1.1.0             Common String Manipulations with regard to casing and...
doctrine/instantiator                  1.2.0              A small, lightweight utility to instantiate objects i...
dotpointer/mysql-shim                  master master     
facebook/webdriver                     1.6.0              A PHP client for Selenium WebDriver
fzaninotto/faker                       dev-master 65fbcca Faker is a PHP library that generates fake data for you.
grasmash/expander                      1.0.0              Expands internal property references in PHP arrays file.
grasmash/yaml-expander                 1.4.0              Expands internal property references in a yaml file.
guzzlehttp/guzzle                      6.3.3              Guzzle is a PHP HTTP client library
guzzlehttp/promises                    v1.3.1             Guzzle promises library
guzzlehttp/psr7                        1.5.2              PSR-7 message implementation that also provides commo...
hamcrest/hamcrest-php                  v1.2.2             This is the PHP port of Hamcrest Matchers
hautelook/phpass                       dev-master c3ea12e Portable PHP password hashing framework
hoa/consistency                        1.17.05.02         The Hoa\Consistency library.
hoa/console                            3.17.05.02         The Hoa\Console library.
hoa/event                              1.17.01.13         The Hoa\Event library.
hoa/exception                          1.17.01.16         The Hoa\Exception library.
hoa/file                               1.17.07.11         The Hoa\File library.
hoa/iterator                           2.17.01.10         The Hoa\Iterator library.
hoa/protocol                           1.17.01.14         The Hoa\Protocol library.
hoa/stream                             1.17.02.21         The Hoa\Stream library.
hoa/ustring                            4.17.01.16         The Hoa\Ustring library.
illuminate/contracts                   v5.3.23            The Illuminate Contracts package.
illuminate/support                     v5.3.23            The Illuminate Support package.
league/container                       2.4.1              A fast and intuitive dependency injection container.
league/csv                             6.3.0              Csv data manipulation made easy in PHP
mikecao/flight                         v1.1               Flight is a fast, simple, extensible framework for PH...
mikey179/vfsStream                     v1.6.5             Virtual file system to mock the real file system in u...
mockery/mockery                        0.9.9              Mockery is a simple yet flexible PHP mock object fram...
monolog/monolog                        1.23.0             Sends your logs to files, sockets, inboxes, databases...
mustache/mustache                      v2.12.0            A Mustache implementation in PHP.
myclabs/deep-copy                      1.9.1              Create deep copies (clones) of your objects
netcarver/textile                      v3.5.1             A lightweight markup language that takes (almost) pla...
paragonie/random_compat                v2.0.12            PHP 5.x polyfill for random_bytes() and random_int() ...
patchwork/utf8                         v1.3.1             Portable and performant UTF-8, Unicode and Grapheme C...
paulbunyannet/autoversion              1.2.0              Auto versioning class for front end assets.
paulbunyannet/bandolier                1.6.5              Reusable utilities
paulbunyannet/codeception-helpers      dev-master db309fb Helpers for Codeception tests.
phar-io/manifest                       1.0.3              Component for reading phar.io manifest information fr...
phar-io/version                        2.0.1              Library for handling version information and constraints
php-activerecord/php-activerecord      v1.2.0             php-activerecord is an open source ORM library based ...
phpdocumentor/reflection-common        1.0.1              Common reflection classes used by phpdocumentor to re...
phpdocumentor/reflection-docblock      4.3.1              With this component, a library can provide support fo...
phpdocumentor/type-resolver            0.4.0             
phpspec/prophecy                       1.8.0              Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage              7.0.3              Library that provides collection, processing, and ren...
phpunit/php-file-iterator              2.0.2              FilterIterator implementation that filters files base...
phpunit/php-text-template              1.2.1              Simple template engine.
phpunit/php-timer                      2.1.1              Utility class for timing
phpunit/php-token-stream               3.0.1              Wrapper around PHP's tokenizer extension.
phpunit/phpunit                        8.1.5              The PHP Unit Testing framework.
pimple/pimple                          v2.1.1             Pimple is a simple Dependency Injection Container for...
pixel418/markdownify                   v2.1.6             The HTML to Markdown converter for PHP 
psr/cache                              1.0.1              Common interface for caching libraries
psr/container                          1.0.0              Common Container Interface (PHP FIG PSR-11)
psr/http-message                       1.0.1              Common interface for HTTP messages
psr/log                                1.1.0              Common interface for logging libraries
ralouphie/getallheaders                2.0.5              A polyfill for getallheaders.
rcastera/session                       v1.0.0             Session Class
react/promise                          dev-master c1aad8e A lightweight implementation of CommonJS Promises/A f...
respect/validation                     1.1.29             The most awesome validation engine ever created for PHP
robmorgan/phinx                        v0.8.1             Phinx makes it ridiculously easy to manage the databa...
rollbar/rollbar                        v1.7.2             Monitors errors and exceptions and reports them to Ro...
sebastian/code-unit-reverse-lookup     1.0.1              Looks up which function or method a line of code belo...
sebastian/comparator                   3.0.2              Provides the functionality to compare PHP values for ...
sebastian/diff                         3.0.2              Diff implementation
sebastian/environment                  4.2.2              Provides functionality to handle HHVM/PHP environments
sebastian/exporter                     3.1.0              Provides the functionality to export PHP variables fo...
sebastian/global-state                 3.0.0              Snapshotting of global state
sebastian/object-enumerator            3.0.3              Traverses array structures and object graphs to enume...
sebastian/object-reflector             1.1.1              Allows reflection of object attributes, including inh...
sebastian/recursion-context            3.0.0              Provides functionality to recursively process PHP var...
sebastian/resource-operations          2.0.1              Provides a list of PHP built-in functions that operat...
sebastian/version                      2.0.1              Library that helps with managing the version number o...
swiftmailer/swiftmailer                v5.4.9             Swiftmailer, free feature-rich PHP mailer
symfony/browser-kit                    v4.2.8             Symfony BrowserKit Component
symfony/config                         v3.4.9             Symfony Config Component
symfony/console                        v3.4.27            Symfony Console Component
symfony/contracts                      v1.1.0             A set of abstractions extracted out of the Symfony co...
symfony/css-selector                   v4.2.8             Symfony CssSelector Component
symfony/debug                          v4.2.8             Symfony Debug Component
symfony/dom-crawler                    v4.2.8             Symfony DomCrawler Component
symfony/event-dispatcher               v4.2.8             Symfony EventDispatcher Component
symfony/filesystem                     v4.2.8             Symfony Filesystem Component
symfony/finder                         v4.2.8             Symfony Finder Component
symfony/polyfill-ctype                 v1.11.0            Symfony polyfill for ctype functions
symfony/polyfill-mbstring              v1.11.0            Symfony polyfill for the Mbstring extension
symfony/process                        v4.2.8             Symfony Process Component
symfony/yaml                           v3.4.9             Symfony Yaml Component
tedivm/stash                           v0.15.1            The place to keep your cache.
theseer/tokenizer                      1.1.2              A small library for converting tokenized PHP source c...
twig/twig                              v1.15.1            Twig, the flexible, fast, and secure template languag...
vlucas/phpdotenv                       v2.4.0             Loads environment variables from `.env` to `getenv()`...
webmozart/assert                       1.4.0              Assertions to validate method input/output with nice ...
whatdafox/codeception-mailtrap         1.2.5              Mailtrap module for Codeception
withremote/rah_flat                    master master     anahkiasen/underscore-php              2.0.0              A redacted port of Underscore.js for PHP
behat/gherkin                          v4.6.0             Gherkin DSL parser for PHP 5.3
brainsware/php-markdown-extra-extended 0.1.0              PHP Markdown & Extra
brandonwamboldt/utilphp                1.0.5              util.php is a collection of useful functions and snip...
codeception/c3                         2.4.1              CodeCoverage collector for Codeception
codeception/codeception                3.0.1              BDD-style testing framework
codeception/phpunit-wrapper            8.0.4              PHPUnit classes used by Codeception
codeception/robo-paracept              0.4.2              Codeception Parallel Execution Tasks via Robo Task Ru...
codeception/stub                       2.1.0              Flexible Stub wrapper for PHPUnit's Mock Builder
codeigniter/framework                  3.1.10             The CodeIgniter framework
consolidation/annotated-command        2.12.0             Initialize Symfony Console commands from annotated co...
consolidation/config                   1.2.1              Provide configuration services for a commandline tool.
consolidation/log                      1.1.1              Improved Psr-3 / Psr\Log logger based on Symfony Cons...
consolidation/output-formatters        3.4.1              Format text by applying transformations provided by p...
consolidation/robo                     1.4.9              Modern task runner
consolidation/self-update              1.1.5              Provides a self:update command for Symfony Console ap...
container-interop/container-interop    1.2.0              Promoting the interoperability of container objects (...
crazyinventor/mailtrapper              master master     
dflydev/dot-access-data                v1.1.0             Given a deep data structure, access data by dot notat...
doctrine/inflector                     v1.1.0             Common String Manipulations with regard to casing and...
doctrine/instantiator                  1.2.0              A small, lightweight utility to instantiate objects i...
dotpointer/mysql-shim                  master master     
facebook/webdriver                     1.6.0              A PHP client for Selenium WebDriver
fzaninotto/faker                       dev-master 65fbcca Faker is a PHP library that generates fake data for you.
grasmash/expander                      1.0.0              Expands internal property references in PHP arrays file.
grasmash/yaml-expander                 1.4.0              Expands internal property references in a yaml file.
guzzlehttp/guzzle                      6.3.3              Guzzle is a PHP HTTP client library
guzzlehttp/promises                    v1.3.1             Guzzle promises library
guzzlehttp/psr7                        1.5.2              PSR-7 message implementation that also provides commo...
hamcrest/hamcrest-php                  v1.2.2             This is the PHP port of Hamcrest Matchers
hautelook/phpass                       dev-master c3ea12e Portable PHP password hashing framework
hoa/consistency                        1.17.05.02         The Hoa\Consistency library.
hoa/console                            3.17.05.02         The Hoa\Console library.
hoa/event                              1.17.01.13         The Hoa\Event library.
hoa/exception                          1.17.01.16         The Hoa\Exception library.
hoa/file                               1.17.07.11         The Hoa\File library.
hoa/iterator                           2.17.01.10         The Hoa\Iterator library.
hoa/protocol                           1.17.01.14         The Hoa\Protocol library.
hoa/stream                             1.17.02.21         The Hoa\Stream library.
hoa/ustring                            4.17.01.16         The Hoa\Ustring library.
illuminate/contracts                   v5.3.23            The Illuminate Contracts package.
illuminate/support                     v5.3.23            The Illuminate Support package.
league/container                       2.4.1              A fast and intuitive dependency injection container.
league/csv                             6.3.0              Csv data manipulation made easy in PHP
mikecao/flight                         v1.1               Flight is a fast, simple, extensible framework for PH...
mikey179/vfsStream                     v1.6.5             Virtual file system to mock the real file system in u...
mockery/mockery                        0.9.9              Mockery is a simple yet flexible PHP mock object fram...
monolog/monolog                        1.23.0             Sends your logs to files, sockets, inboxes, databases...
mustache/mustache                      v2.12.0            A Mustache implementation in PHP.
myclabs/deep-copy                      1.9.1              Create deep copies (clones) of your objects
netcarver/textile                      v3.5.1             A lightweight markup language that takes (almost) pla...
paragonie/random_compat                v2.0.12            PHP 5.x polyfill for random_bytes() and random_int() ...
patchwork/utf8                         v1.3.1             Portable and performant UTF-8, Unicode and Grapheme C...
paulbunyannet/autoversion              1.2.0              Auto versioning class for front end assets.
paulbunyannet/bandolier                1.6.5              Reusable utilities
paulbunyannet/codeception-helpers      dev-master db309fb Helpers for Codeception tests.
phar-io/manifest                       1.0.3              Component for reading phar.io manifest information fr...
phar-io/version                        2.0.1              Library for handling version information and constraints
php-activerecord/php-activerecord      v1.2.0             php-activerecord is an open source ORM library based ...
phpdocumentor/reflection-common        1.0.1              Common reflection classes used by phpdocumentor to re...
phpdocumentor/reflection-docblock      4.3.1              With this component, a library can provide support fo...
phpdocumentor/type-resolver            0.4.0             
phpspec/prophecy                       1.8.0              Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage              7.0.3              Library that provides collection, processing, and ren...
phpunit/php-file-iterator              2.0.2              FilterIterator implementation that filters files base...
phpunit/php-text-template              1.2.1              Simple template engine.
phpunit/php-timer                      2.1.1              Utility class for timing
phpunit/php-token-stream               3.0.1              Wrapper around PHP's tokenizer extension.
phpunit/phpunit                        8.1.5              The PHP Unit Testing framework.
pimple/pimple                          v2.1.1             Pimple is a simple Dependency Injection Container for...
pixel418/markdownify                   v2.1.6             The HTML to Markdown converter for PHP 
psr/cache                              1.0.1              Common interface for caching libraries
psr/container                          1.0.0              Common Container Interface (PHP FIG PSR-11)
psr/http-message                       1.0.1              Common interface for HTTP messages
psr/log                                1.1.0              Common interface for logging libraries
ralouphie/getallheaders                2.0.5              A polyfill for getallheaders.
rcastera/session                       v1.0.0             Session Class
react/promise                          dev-master c1aad8e A lightweight implementation of CommonJS Promises/A f...
respect/validation                     1.1.29             The most awesome validation engine ever created for PHP
robmorgan/phinx                        v0.8.1             Phinx makes it ridiculously easy to manage the databa...
rollbar/rollbar                        v1.7.2             Monitors errors and exceptions and reports them to Ro...
sebastian/code-unit-reverse-lookup     1.0.1              Looks up which function or method a line of code belo...
sebastian/comparator                   3.0.2              Provides the functionality to compare PHP values for ...
sebastian/diff                         3.0.2              Diff implementation
sebastian/environment                  4.2.2              Provides functionality to handle HHVM/PHP environments
sebastian/exporter                     3.1.0              Provides the functionality to export PHP variables fo...
sebastian/global-state                 3.0.0              Snapshotting of global state
sebastian/object-enumerator            3.0.3              Traverses array structures and object graphs to enume...
sebastian/object-reflector             1.1.1              Allows reflection of object attributes, including inh...
sebastian/recursion-context            3.0.0              Provides functionality to recursively process PHP var...
sebastian/resource-operations          2.0.1              Provides a list of PHP built-in functions that operat...
sebastian/version                      2.0.1              Library that helps with managing the version number o...
swiftmailer/swiftmailer                v5.4.9             Swiftmailer, free feature-rich PHP mailer
symfony/browser-kit                    v4.2.8             Symfony BrowserKit Component
symfony/config                         v3.4.9             Symfony Config Component
symfony/console                        v3.4.27            Symfony Console Component
symfony/contracts                      v1.1.0             A set of abstractions extracted out of the Symfony co...
symfony/css-selector                   v4.2.8             Symfony CssSelector Component
symfony/debug                          v4.2.8             Symfony Debug Component
symfony/dom-crawler                    v4.2.8             Symfony DomCrawler Component
symfony/event-dispatcher               v4.2.8             Symfony EventDispatcher Component
symfony/filesystem                     v4.2.8             Symfony Filesystem Component
symfony/finder                         v4.2.8             Symfony Finder Component
symfony/polyfill-ctype                 v1.11.0            Symfony polyfill for ctype functions
symfony/polyfill-mbstring              v1.11.0            Symfony polyfill for the Mbstring extension
symfony/process                        v4.2.8             Symfony Process Component
symfony/yaml                           v3.4.9             Symfony Yaml Component
tedivm/stash                           v0.15.1            The place to keep your cache.
theseer/tokenizer                      1.1.2              A small library for converting tokenized PHP source c...
twig/twig                              v1.15.1            Twig, the flexible, fast, and secure template languag...
vlucas/phpdotenv                       v2.4.0             Loads environment variables from `.env` to `getenv()`...
webmozart/assert                       1.4.0              Assertions to validate method input/output with nice ...
whatdafox/codeception-mailtrap         1.2.5              Mailtrap module for Codeception
withremote/rah_flat                    master master     
```
* Suite configuration:

```yml
# Codeception Test Suite Configuration

# suite for unit (internal) tests.
class_name: UnitTester
modules:
  enabled:
      - Asserts
      - \Helper\Unit
      - Db:
          cleanup: true
```

**Comment by Naktibalda** [comment]
@DavertMik can you finally explain why does it ever work?
https://github.com/Codeception/Codeception/blob/3.0/src/Codeception/Test/Test.php#L67 has no return type declaration,
https://github.com/sebastianbergmann/phpunit/blob/master/src/Framework/Test.php#L22 has return type declaration since 7.0.0

**Comment by DavertMik** [comment]
I think the `codeception/phpunit-wrapper` files were not initialized.

You should execute  `Codeception\PHPUnit\Init::init();` before running everything else.

> @DavertMik can you finally explain why does it ever work?
https://github.com/Codeception/Codeception/blob/3.0/src/Codeception/Test/Test.php#L67 has no return type declaration,
https://github.com/sebastianbergmann/phpunit/blob/master/src/Framework/Test.php#L22 has return type declaration since 7.0.0

Here is overridden interface https://github.com/Codeception/phpunit-wrapper/blob/8.0/src/phpunit7-interfaces.php#L8

**Comment by natenolting** [comment]
> I think the `codeception/phpunit-wrapper` files were not initialized.
> 
> You should execute `Codeception\PHPUnit\Init::init();` before running everything else.

@DavertMik Where would `Codeception\PHPUnit\Init::init();` go? In the `_before()` step?

**Comment by svycka** [comment]
@Naktibalda why did you close this? at least explain how to fix it.
@natenolting have you able to fix this?

**Comment by svycka** [comment]
okay, it seems the order of installing Codeception PHPUnit does matter. You have to install PHPUnit and only then Codeception to make it work. But I don't feel save to add this in my  deployment.

**Comment by natenolting** [comment]
> @Naktibalda why did you close this? at least explain how to fix it.
> @natenolting have you able to fix this?
@svycka No I was not, I was waiting on some feedback.

**Comment by Naktibalda** [comment]
@svycka Probably I thought that the question was answered by @DavertMik 

There is a relevant comment in another ticket: https://github.com/Codeception/Codeception/issues/5031#issuecomment-507222146

There it was identified that the issue happened because project had multiple dependencies implementing compatibility with different versions of PHPUnit.

**Comment by svycka** [comment]
@Naktibalda maybe reopen this still happens in 4.0 any workaround?

**Comment by Naktibalda** [comment]
@svycka This is the same issue as #5031 

Probably you have another dependency which implements compability with multiple major versions of PHPUnit and it conflicts with Codeception's compatibility library.

Unless you want to investigate and fix it, this issue will disappear when some major version of PHPUnit is so incompatible that it requires a new major version of Codeception to support it.
It could happen this February or the next one.

A workaround is to choose between Codeception and conflicting library,
or use an older version of PHPUnit if there is a version which is compatible with all dependencies.

**Comment by svycka** [comment]
@Naktibalda you are right it conflicts with `laminas/laminas-test`. 
Will new Codeception version drop legacy dependencies?
